enhancements/installer/pre-flight-validations.md

```diff
@@ -0,0 +1,168 @@
+
+However, this doesn't check that the environment is suitable to install OpenShift:
+- pull secret is valid to fetch the container images
+- the tenant has adequate quota and the flavors' specifications are within the
+  recommended ranges.
+- for user-provided networks, check the subnets have a DHCP server and valid
+  DNS to reach the cloud's endpoints
+- required cloud services are available
+- required storage performance
+
+The pre-flight validations should not alter the target infrastructure nor leave
+behind any new resource.
+
+### Non-Goals
+
+Implementing every useful validation we can think of is out-of-scope:
+- The potential number of validations is huge
+- We expect more validations to be added over time as new issues are
+  discovered.
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### User Stories
+
+#### On-premise deployments
+
+As an OpenShift administrator installing a new OpenShift cluster, I want the
+installation process to fail early when the requirements are not met for a
+successful installation. In such a case, I also want clear and actionable error
+messages right in the Installer output.
+
+#### CI debugging
+
+As an OpenShift developer, I would like to rapidly identify failures caused by
+transient environmental issues.
+
+### Implementation Details/Notes/Constraints
+
+The validations must leave the environment unaltered.
+
+The framework should allow implementing checks common to all platforms as well
+as per-platform checks.
+
+We do not envision the need for the users to write additional validations. As
+a consequence, the validations do not need to be loaded on startup and will be
+compiled into the `openshift-install` binary. This may be revisited later.
+
+#### Enabling the validations
+
+The validations will run automatically, right after the `install-config.yaml`
+syntax validation.
+
+#### Reporting errors
+
+A failed validation typically causes the installer to fail early and not
+proceed with the deployment of OpenShift.
+
+The installer will report all found failures at once, and will not stop on the
+first validation error.
+
+Validation failures should clearly indicate the root cause of an error, and if
+applicable, suggest a solution to it.
+
+### Risks and Mitigations
+
+Depending on the number and the nature of the validations, the installation
+time might end up increasing noticeably. Because the goal is to fail early,
+time-consuming validations should seldom be considered for inclusion.
+
+Since pre-flight validations are only run at install time, and not on cluster
+upgrade/downgrade, caution should be used when migrating existing validations
+to this "pre-flight" framework.
+
+## Design Details
+
+### Test Plan
+
+The code for the framework and the validations will have unit tests.
+
+In addition, we will enable the validations checks in CI in order to exercise
+them and potentially highlight issues with the underlying CI infrastructure.
```

Comment:
It would be great to be able to run pre-flight validations separately, for 2 use-cases:

* in CI, we would disable pre-check validations everywhere (most likely) if they take too much time from the CI job and can cause timeout issues (it happened a lot in OpenStack CI).
* a customer who want to first check their infrastructure, then run the install

Therefore I propose that we have a subcommand like ```openshift-install validate pre-flight``` for example.